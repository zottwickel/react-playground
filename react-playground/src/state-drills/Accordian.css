.listContainer {
    width: 75%;
    margin-left: auto;
    margin-right: auto;
    list-style-type: none;
    border: 2px solid black;
    padding: 0;
}

li {
    width: 100%;
    margin: 0;
    padding: 0;
}

button {
    font-size: 20px;
    color: white;
    background-color: black;
    border: 2px solid gray;
    width: 100%;
    margin: 0;
}

p {
    background-color: silver;
    width: 100%;
    margin: 0;
}