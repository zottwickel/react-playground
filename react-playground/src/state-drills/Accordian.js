import React from 'react';
import './Accordian.css';

class Accordian extends React.Component {
    static defaultProps = { sections: [ { title: false } ] }
    state = {
        activeSection: null
    };

    handleButtonClick(i) {
        this.setState({ activeSection: i })
    }

    renderContent() {
        const currentSection = this.props.sections[this.state.activeSection]
        return (
            <p>{currentSection.content}</p>
        )
    }
    
    renderButton(section, i) {
        return (
            <button onClick={() => this.handleButtonClick(i)}>
                {section.title}
            </button>
        )
    }

    renderList() {
        return this.props.sections.map((section, i) => (
            <li key={i}>
                {!!this.props.sections[0].title && this.renderButton(section, i)}
                {this.state.activeSection === i && this.renderContent()}
            </li>
        ))
    }

    render() {
        return (
            <ul className='listContainer'>
                {this.renderList()}
            </ul>
        )
    }
}

export default Accordian;