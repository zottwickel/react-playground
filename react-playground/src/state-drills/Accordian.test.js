import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Accordian from './Accordian';

describe('Accordian component', () => {
    const sectionsProp = [
        {
            title: 'Section 1',
            content: 'Lorem ipsum dolor sit amet consectetur adipisicing elit.',
        },
        {
            title: 'Section 2',
            content: 'Cupiditate tenetur aliquam necessitatibus id distinctio quas nihil ipsam nisi modi!',
        },
        {
            title: 'Section 3',
            content: 'Animi amet cumque sint cupiditate officia ab voluptatibus libero optio et?',
        },
    ]
    it('renders without errors', () => {
        const div = document.createElement('div');
        ReactDOM.render(<Accordian />, div);
        ReactDOM.unmountComponentAtNode(div);
    })

    it('renders an empty li when the sections prop is not supplied', () => {
        const wrapper = shallow(<Accordian />)
        expect(toJson(wrapper)).toMatchSnapshot()
    })

    it('renders no sections active by default', () => {
        const wrapper = shallow(<Accordian sections={sectionsProp} />)
        expect(toJson(wrapper)).toMatchSnapshot()
    })

    it('opens a clicked section', () => {
        const wrapper = shallow(<Accordian sections={sectionsProp} />)
        wrapper.find('button').at(0).simulate('click')
        expect(toJson(wrapper)).toMatchSnapshot()
    })

    it('only opens one section at a time', () => {
        const wrapper = shallow(<Accordian sections={sectionsProp} />)
        wrapper.find('button').at(0).simulate('click')
        wrapper.find('button').at(1).simulate('click')
        expect(toJson(wrapper)).toMatchSnapshot()
    })
    
})