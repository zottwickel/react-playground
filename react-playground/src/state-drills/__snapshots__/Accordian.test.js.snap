// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordian component only opens one section at a time 1`] = `
<ul>
  <li
    key="0"
  >
    <button
      onClick={[Function]}
    >
      Section 1
    </button>
  </li>
  <li
    key="1"
  >
    <button
      onClick={[Function]}
    >
      Section 2
    </button>
    <p>
      Cupiditate tenetur aliquam necessitatibus id distinctio quas nihil ipsam nisi modi!
    </p>
  </li>
  <li
    key="2"
  >
    <button
      onClick={[Function]}
    >
      Section 3
    </button>
  </li>
</ul>
`;

exports[`Accordian component opens a clicked section 1`] = `
<ul>
  <li
    key="0"
  >
    <button
      onClick={[Function]}
    >
      Section 1
    </button>
    <p>
      Lorem ipsum dolor sit amet consectetur adipisicing elit.
    </p>
  </li>
  <li
    key="1"
  >
    <button
      onClick={[Function]}
    >
      Section 2
    </button>
  </li>
  <li
    key="2"
  >
    <button
      onClick={[Function]}
    >
      Section 3
    </button>
  </li>
</ul>
`;

exports[`Accordian component renders an empty li when the sections prop is not supplied 1`] = `
<ul>
  <li
    key="0"
  />
</ul>
`;

exports[`Accordian component renders no sections active by default 1`] = `
<ul>
  <li
    key="0"
  >
    <button
      onClick={[Function]}
    >
      Section 1
    </button>
  </li>
  <li
    key="1"
  >
    <button
      onClick={[Function]}
    >
      Section 2
    </button>
  </li>
  <li
    key="2"
  >
    <button
      onClick={[Function]}
    >
      Section 3
    </button>
  </li>
</ul>
`;
